java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java

```diff
@@ -747,8 +752,10 @@ static TruthValue evaluatePredicateRange(PredicateLeaf predicate,
                                            ValueRange range,
                                            BloomFilter bloomFilter,
                                            boolean useUTCTimestamp) {
+    // If range is invalid, that means that no value (including null) is written to this column
+    // we should return TruthValue.NO for any predicate.
     if (!range.isValid()) {
-      return TruthValue.YES_NO_NULL;
+      return TruthValue.NO;
```

Comment:
I'm not sure if this is safe. IMHO, we'd better to be conservative by returning all data if we're uncertain about any filter.

Reply:
@wgtmac , Thanks for the review, but I think the change here is the key to fixing the bug.
The schema in the unit test is `struct<x:struct<z:double>,y:int>`
I've constructed 1024 rows of data, and `x` is null. This means that the statistics for `x.z` are kept in the initial state `numberOfValues = 0 hasNull = false`. 

Any filtering of `x.z` at this point should return `TruthValue.NO`

Here's the enumeration operator: for data that doesn't exist (which distinguishes it from having a value and being Null) I think it's valid to return `TruthValue.NO`. If there is a counterexample, I'd be happy to modify it.

```java
public interface PredicateLeaf {

  /**
   * The possible operators for predicates. To get the opposites, construct
   * an expression with a not operator.
   */
  public static enum Operator {
    EQUALS,
    NULL_SAFE_EQUALS,
    LESS_THAN,
    LESS_THAN_EQUALS,
    IN,
    BETWEEN,
    IS_NULL
  }
  ....
}
```

Reply:
I will check if it helped in my situation.

But am also concerned about this particular line.

As I understand https://issues.apache.org/jira/browse/ORC-1075, this line was **correct**.

Reply:
This pr uses the following judgment instead of ORC-1075, which is more accurate.
```java
if (statsProto == EMPTY_COLUMN_STATISTICS) {
    return TruthValue.YES_NO_NULL;
}
```

Reply:
Right!